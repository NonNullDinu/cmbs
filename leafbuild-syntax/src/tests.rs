fn parses_without_errors(s: &str) {
    assert_eq!(crate::parser::parse(s).errors, vec![]);
}

// Those are only *parsing* tests, so as long as they make sense
// *syntactically*, they should parse.
// Even if semantically they are void of meaning.
#[test]
fn unit_to_unit_assignment() {
    parses_without_errors("() = ()\n")
}

#[test]
fn line_trailing_space() {
    parses_without_errors("() = () \n")
}

#[test]
fn if_condition() {
    parses_without_errors("if (()) {} else {}")
}

#[test]
fn if_else_if_condition() {
    parses_without_errors("if () {} else if () {} else {}")
}

#[test]
fn if_else_if_else_condition_stretched() {
    parses_without_errors(
        r#"
    if ()

    {



    }

    else if () {


    }

    else {


    }
    "#,
    )
}

#[test]
fn if_condition_in_expr() {
    parses_without_errors("() = if () {} else {} \n")
}

#[test]
fn if_condition_in_expr_stretched() {
    parses_without_errors(
        r#"
    () = if ()
    {

    }

    else
    {


    }
    "#,
    )
}

#[test]
fn declaration() {
    parses_without_errors("let x = ()\n")
}

#[test]
fn foreach_basic() {
    parses_without_errors("foreach () in () {}")
}
