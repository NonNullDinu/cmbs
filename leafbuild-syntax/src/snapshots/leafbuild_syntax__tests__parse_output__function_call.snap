---
source: leafbuild-syntax/src/tests.rs
expression: "\n        x = f(1, 2, a = b)\n    "

---
- ROOT 0..32
  - "\n" NEWLINE 0..1
  - "        " WHITESPACE 1..9
  - Assignment 9..28
    - Expr 9..10
      - PrimaryExpr 9..10
        - "x" ID 9..10
    - " " WHITESPACE 10..11
    - "=" EQ 11..12
    - " " WHITESPACE 12..13
    - Expr 13..27
      - FuncCallExpr 13..27
        - PrimaryExpr 13..14
          - "f" ID 13..14
        - FuncCallArgs 14..27
          - "(" L_PAREN 14..15
          - Expr 15..16
            - PrimaryExpr 15..16
              - "1" NUMBER 15..16
          - "," COMMA 16..17
          - " " WHITESPACE 17..18
          - Expr 18..19
            - PrimaryExpr 18..19
              - "2" NUMBER 18..19
          - "," COMMA 19..20
          - " " WHITESPACE 20..21
          - KExpr 21..26
            - "a" ID 21..22
            - " " WHITESPACE 22..23
            - "=" EQ 23..24
            - " " WHITESPACE 24..25
            - Expr 25..26
              - PrimaryExpr 25..26
                - "b" ID 25..26
          - ")" R_PAREN 26..27
    - "\n" NEWLINE 27..28
  - "    " WHITESPACE 28..32

